### PR TITLE
(PUP-10485) quote pip path when it has spaces

### PR DESCRIPTION
**Pull request by tkishel**
Quoting is required if the path to the pip command contains spaces.
Required for execpipe() but not execute(), and quoting with execute() causes an error.

### CODE REVIEW

**Comment by luchihoratiu** [comment]
LGTM! 👍 

Managed to reproduce the issue and manually tested the proposed fix on Windows 2016. Everything seems to work as expected. Also checked a couple of Linux operating systems and they indeed don't seem to bother having commands quoted. 

Added some unit tests to cover the new cases where quoting is needed (integration tests seemed a bit expensive for these changes but could be done if any concerns arise). The two commits should be squashed together **before merging** the pull request (easier to review this way than directly amending to @tkishel's work) and I would personally **change the title** to something like `Quotes pip path when it has spaces` since the code quotes the whole path not just the spaces as the current title suggests. 

While digging through the code I also found the reason why this change is needed only for `execpipe` and creates errors for `execute`. The `execute` method already checks if it's running on Windows and quotes the command if it contains spaces but for our case it would over do it (we end up getting double quotes such as `""path""`) and cause errors. I initially wanted to suggest solving the issue from its roots (util/execution.rb) by doing the same for `execpipe` but sadly that would result in a bigger than necessary refactoring because `execute` has the `command` and its `arguments/options` passed to the method separately (safe to do changes to the command/s when needed) while `execpipe` has them all together in the same array (seems a bit regression prone as pip is not the only one using this method).

Since this kind of paths are common to come across only on Windows (because of default system folder structuring at least) and, out of all of the current `execpipe` method callers (_provider/package/dpkg.rb, provider/package/openbsd.rb, provider/package/opkg.rb, provider/package/pacman.rb, provider/package/pip.rb, provider/package/rpm.rb, provider/selboolean/getsetsebool.rb, provider/selmodule/semodule.rb, provider/service/openbsd.rb, provider/service/upstart.rb_), the `pip package provider` is the only one in danger to come across this issue, I think that this solution is best as it is for now.

File: lib/puppet/provider/package/pip.rb [hunk 1]
```diff
@@ -227,12 +227,14 @@ def install_options
     join_options(@resource[:install_options])
   end
 
+  # Quoting is required if the path to the pip command contains spaces.
+  # Required for execpipe() but not execute(), and quoting with execute() causes an error.
```
Comment (luchihoratiu) [inline lib/puppet/provider/package/pip.rb:231]: nit: I would replace `and` with `because`. Makes more sense when reading (for me at least).

Comment (tkishel) [reply lib/puppet/provider/package/pip.rb]: Done and force pushed.

**Comment by tkishel** [comment]
That's great research. Possibly another option would be to standardize on `execute` instead of `execpipe` but that is too large of a scope for now.

**Pull request merged**